File: extra_1/app/src/main/java/com/example/extra_1/MainActivity.java
package com.example.extra_1;

import android.os.Bundle;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    private TextView currentTemp;
    private TextView currentDescription;
    private ForecastAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        currentTemp = findViewById(R.id.currentTemp);
        currentDescription = findViewById(R.id.currentDescription);
        RecyclerView weeklyRecycler = findViewById(R.id.weeklyRecycler);
        weeklyRecycler.setLayoutManager(new LinearLayoutManager(this));
        adapter = new ForecastAdapter();
        weeklyRecycler.setAdapter(adapter);

        WeatherViewModel viewModel = new ViewModelProvider(this).get(WeatherViewModel.class);
        viewModel.getTodayWeather().observe(this, this::renderToday);
        viewModel.getWeekWeather().observe(this, this::renderWeek);
        viewModel.getError().observe(this, error -> {
            if (error != null) {
                currentDescription.setText(error);
            }
        });

        viewModel.loadWeather("Moscow");
    }

    private void renderToday(WeatherInfo info) {
        if (info == null) return;
        currentTemp.setText(String.format("%.1f°C", info.getTempCelsius()));
        currentDescription.setText(info.getDescription());
    }

    private void renderWeek(List<WeatherInfo> forecast) {
        if (forecast == null) return;
        adapter.submitList(forecast);
    }
}

File: extra1/app/src/main/java/com/example/extra1/MainActivity.kt
package com.example.extra1

import android.os.Bundle
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.extra1.ui.ForecastAdapter
import com.example.extra1.ui.WeatherViewModel

class MainActivity : AppCompatActivity() {
    private val vm: WeatherViewModel by viewModels()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val today = findViewById<TextView>(R.id.todayWeather)
        val adapter = ForecastAdapter()
        findViewById<RecyclerView>(R.id.weekRecycler).apply {
            layoutManager = LinearLayoutManager(this@MainActivity)
            this.adapter = adapter
        }

        vm.today.observe(this) { today.text = it }
        vm.week.observe(this) { adapter.submit(it) }
        vm.load(city = "Almaty", apiKey = "PUT_OPEN_WEATHER_MAP_KEY_HERE")
    }
}
